sdk/messaging/azservicebus/receiver.go

```diff
@@ -333,118 +348,114 @@ func (r *Receiver) receiveMessagesImpl(ctx context.Context, maxMessages int, opt
 	//    user isn't actually waiting for anymore. So we make sure that #3 runs if the
 	//    link is still valid.
 	// Phase 3. <drain the link and leave it in a good state>
-	_, receiver, _, linksRevision, err := r.amqpLinks.Get(ctx)
 
-	if err != nil {
-		if err := r.amqpLinks.RecoverIfNeeded(ctx, linksRevision, err); err != nil {
-			return nil, err
-		}
-
-		return nil, err
-	}
-
-	if err := receiver.IssueCredit(uint32(maxMessages)); err != nil {
-		_ = r.amqpLinks.RecoverIfNeeded(ctx, linksRevision, err)
-		return nil, err
-	}
+	var all []*ReceivedMessage
+	var cancel context.CancelFunc
+	needsDrain := true
 
 	defaultTimeAfterFirstMessage := 20 * time.Millisecond
 
 	if r.receiveMode == ReceiveModeReceiveAndDelete {
 		defaultTimeAfterFirstMessage = time.Second
 	}
 
-	messages, err := r.getMessages(ctx, receiver, maxMessages, defaultTimeAfterFirstMessage)
+	var linksWithID *internal.LinksWithID
+
+	err := r.amqpLinks.Retry(ctx, "receiveMessages", func(ctx context.Context, lwid *internal.LinksWithID, args *utils.RetryFnArgs) error {
```

Comment:
@jhendrixMSFT: the changes in here are a bit more meaty than other spots. 

Basically, I just wanted to make sure that the retry() loop existed around the entire receive operation. Before this it was only around certain spots, which made it harder to just read through (I think). I also inlined some code to make it more obvious what is happening.